v0/util/index.js

```diff
@@ -1474,6 +1479,110 @@ function getValidDynamicFormConfig(
+ * @param {*} error - error occurred during transformation of a single rudder event
+ * @param {string} destType - destination name
+ * @returns
+ */
+const handleRtTfSingleEventError = (input, error, destType) => {
+  const errObj = generateErrorObject(
+    error,
+    destType, // Preference is destination definition name
+    TRANSFORMER_METRIC.TRANSFORMER_STAGE.TRANSFORM
+  );
+  return getErrorRespEvents(
+    [input.metadata],
+    // eslint-disable-next-line no-nested-ternary
+    error.response
+      ? error.response.status
+      : error.code
+      ? error.code
+      : error.status || 400,
+    error.message || "Error occurred while processing payload.",
+    errObj.statTags
+  );
+};
+
+/**
+ * This method is used to check if the input events sent to router transformation are valid
+ * It is to be used only for router transform destinations
+ *
+ * Example:
+ * ```
+ * const errorRespEvents = checkInvalidRtTfEvents(inputs, destType);
+ * if (errorRespEvents.length > 0) {
+ *  return errorRespEvents;
+ * }
+ * ```
+ * @param {Array<object>} inputs - list of rudder events to be transformed
+ * @param {String} destType - destination name
+ * @returns {Array<object> | []}
+ */
+const checkInvalidRtTfEvents = (inputs, destType) => {
+  let destTyp = destType;
+  if (!destTyp) {
+    destTyp = "";
+  }
+  if (!Array.isArray(inputs) || inputs.length <= 0) {
+    const respEvents = getErrorRespEvents(null, 400, "Invalid event array", {
+      destType: destTyp.toUpperCase(),
+      stage: TRANSFORMER_METRIC.TRANSFORMER_STAGE.TRANSFORM,
+      scope: TRANSFORMER_METRIC.MEASUREMENT_TYPE.EXCEPTION.SCOPE
+    });
+    return [respEvents];
+  }
+  return [];
+};
+
+/**
+ * This function serves as a simple implementation of router transformation destinations
+ *
+ * <strong>Note</strong>:
+ * Don't use this method when the following are the scenarios
+ *  - When a specific router transformation logic is involved
+ *  - When batching is involved
+ * @param {Array<object>} inputs - List of rudder events to be transformed
+ * @param {String} destType - destination definition name
+ * @param {Function} singleTfFunc - single event transformation function, we'd recommend this to be an async function(always)
+ * @returns
+ */
+const simpleProcessRouterDest = async (inputs, destType, singleTfFunc) => {
+  const errorRespEvents = checkInvalidRtTfEvents(inputs, destType);
+  if (errorRespEvents.length > 0) {
+    return errorRespEvents;
+  }
+
+  const respList = await Promise.all(
+    inputs.map(async input => {
+      try {
+        if (input.message.statusCode) {
+          // already transformed event
+          return getSuccessRespEvents(
+            input.message,
+            [input.metadata],
+            input.destination
+          );
+        }
+        // if not transformed
+        return getSuccessRespEvents(
+          await singleTfFunc(input),
+          [input.metadata],
+          input.destination
+        );
```

Comment:
> Is there a better way to differentiate?

Currently no! To make this change, we'd need to change how the statusCode is being handled in `rudder-server`. At this point I feel it is not needed. If needed, we can think about this a little deeper and implement

Reply:
Makes sense. What about the event metadata?

Reply:
Even for the check to happen from metadata, we'd need to make changes in server such that the information that the transformation happened already

Reply:
If you really see, it's not always true to skip destination transformation for an already transformed event. There can be cases where in we'd need to go ahead with destination transformation for an already transformed payload

Reply:
### Problem

Let's assume that the event failed first-time and it had access_token_1 at that point. Due to some reason, router tried to send event again then the transformation will not kick-in as this event was already transformed. In this case, there is a chance that the `access_token_1` expired and the event would again & again. We need to think through this and come up with a solution.

### Solution-1

The solution that I have in my mind is that we can have a config at destination-level to make sure, if transformation has to happen. Based on that env, we can control it. This would be a lot clean-way of handling things. If we can have this config change at run-time(without having to restart transformer), it would be an additional plus for us! 


### Conclusion
But this is just a thought, we can probably come-up with other more cleaner, easier solutions as well. I think it's best to leave it this way for now and come back a little later with a more round-about solution for this problem

Reply:
Ok. Let's add this to the tech debt.

Reply:
We are tracking the tech-debt [here](https://www.notion.so/rudderstacks/RouterTf-Handling-of-already-transformed-events-647330ec20574ef8a2ead471985842e3)